packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1210 @@
+    while (node !== null) {
+      if (this.#nodes.firstIn(node) !== null) return true;
+      node = this.#nodes.next(node);
+    }
+    return false;
+  }
+
+  getInboundEdgesByType(
+    to: NodeId,
+  ): {|type: TEdgeType | NullEdgeType, from: NodeId|}[] {
+    let edges = [];
+    let node = this.#nodes.head(to);
+    while (node !== null) {
+      let type = this.#nodes.typeOf(node);
+      let edge = this.#nodes.firstIn(node);
+      while (edge !== null) {
+        let from = this.#edges.from(edge);
+        edges.push({from, type});
+        edge = this.#edges.nextIn(edge);
+      }
+      node = this.#nodes.next(node);
+    }
+    return edges;
+  }
+
+  getOutboundEdgesByType(
+    from: NodeId,
+  ): {|type: TEdgeType | NullEdgeType, to: NodeId|}[] {
+    let edges = [];
+    let node = this.#nodes.head(from);
+    while (node !== null) {
+      let type = this.#nodes.typeOf(node);
+      let edge = this.#nodes.firstOut(node);
+      while (edge !== null) {
+        let to = this.#edges.to(edge);
+        edges.push({to, type});
+        edge = this.#edges.nextOut(edge);
+      }
+      node = this.#nodes.next(node);
+    }
+    return edges;
+  }
+
+  /**
+   * Get the list of nodes connected from this node.
+   */
+  getNodeIdsConnectedFrom(
+    from: NodeId,
+    type:
+      | AllEdgeTypes
+      | TEdgeType
+      | NullEdgeType
+      | Array<TEdgeType | NullEdgeType> = 1,
+  ): NodeId[] {
+    let matches = node =>
+      type === ALL_EDGE_TYPES ||
+      (Array.isArray(type)
+        ? type.includes(this.#nodes.typeOf(node))
+        : type === this.#nodes.typeOf(node));
+
+    let nodes = [];
+    let node = this.#nodes.head(from);
+    while (node !== null) {
+      if (matches(node)) {
+        let edge = this.#nodes.firstOut(node);
+        while (edge !== null) {
+          nodes.push(this.#edges.to(edge));
+          edge = this.#edges.nextOut(edge);
+        }
+      }
+      node = this.#nodes.next(node);
+    }
+    return nodes;
+  }
+
+  /**
+   * Get the list of nodes connected to this node.
+   */
+  getNodeIdsConnectedTo(
+    to: NodeId,
+    type:
+      | AllEdgeTypes
+      | TEdgeType
+      | NullEdgeType
+      | Array<TEdgeType | NullEdgeType> = 1,
+  ): NodeId[] {
+    let matches = node =>
+      type === ALL_EDGE_TYPES ||
+      (Array.isArray(type)
+        ? type.includes(this.#nodes.typeOf(node))
+        : type === this.#nodes.typeOf(node));
+
+    let nodes = [];
+    let node = this.#nodes.head(to);
+    while (node !== null) {
+      if (matches(node)) {
+        let edge = this.#nodes.firstIn(node);
+        while (edge !== null) {
+          nodes.push(this.#edges.from(edge));
+          edge = this.#edges.nextIn(edge);
+        }
+      }
+      node = this.#nodes.next(node);
+    }
+    return nodes;
+  }
+
+  inspect(): any {
+    return {
+      nodes: this.#nodes.inspect(),
+      edges: this.#edges.inspect(),
+    };
+  }
+}
+
+/**
+ * `SharedTypeMap` is a hashmap of items,
```

Comment:
The implementation described here was inspired by [This dissection of the v8 Map internals](https://itnext.io/v8-deep-dives-understanding-map-internals-45eb94a183df)